### PR TITLE
KY-93636 | Added Support for Scheduler Resource and Data source for database backup params

### PR DESCRIPTION
**Pull request by VJ-VijayThales**
1. Added resource support for Scheduler
2. Added datasource support for Datasource
3. Added Support for database backup params in scheduler
4. includes docs, examples, sample scripts and UTs

### CODE REVIEW

**Review by anugram — APPROVED** [review]

File: examples/resources/ciphertrust_scheduler/resource.tf [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+provider "ciphertrust" {
+  # Address of the CipherTrust Manager.
+  address = "https://54.159.702.196"
+  # Username for authentication.
+  username = "admin"
+  # Password for authentication.
+  password = "SamplePass@12"
+  bootstrap = "no"
+}
+
```
Comment (om-shrirao) [inline examples/resources/ciphertrust_scheduler/resource.tf:22]: add scp connection creation and utilize its id in resource scheduler

File: internal/provider/scheduler/data_source_scheduler.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	return &dataSourceScheduler{}
+}
+
+type dataSourceScheduler struct {
+	client *common.Client
+}
+
+type DataSourceModelScheduler struct {
+	Filters types.Map                    `tfsdk:"filters"`
+	Scp     []CreateJobConfigParamsTFSDK `tfsdk:"scheduler"`
```
Comment (om-shrirao) [inline internal/provider/scheduler/data_source_scheduler.go:32]: Scp? it should be scheduler or something

File: internal/provider/scheduler/data_source_scheduler.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+										Attributes: map[string]schema.Attribute{
+											"resource_type": schema.StringAttribute{
+												Computed: true,
+											},
+											"resource_query": schema.StringAttribute{
+												Computed: true,
+											},
+										},
+									},
+								},
```
Comment (om-shrirao) [inline internal/provider/scheduler/data_source_scheduler.go:114]: all response params are also taken right?

File: internal/provider/scheduler/data_source_scheduler.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	tflog.Trace(ctx, common.MSG_METHOD_START+"[resource_scheduler.go -> Read]["+id+"]")
+	var state DataSourceModelScheduler
+	req.Config.Get(ctx, &state)
+	var kvs []string
+	for k, v := range state.Filters.Elements() {
+		kv := fmt.Sprintf("%s=%s&", k, v.(types.String).ValueString())
+		kvs = append(kvs, kv)
+	}
+
+	jsonStr, err := d.client.GetAll(ctx, id, common.URL_SCHEDULER_JOB_CONFIGS+"/?"+strings.Join(kvs, "")+"skip=0&limit=10")
```
Comment (om-shrirao) [inline internal/provider/scheduler/data_source_scheduler.go:141]: i am thinking on this from long, should we make limit as -1? in every resource for data source

File: internal/provider/scheduler/data_source_scheduler.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+								ResourceQuery: types.StringValue(resourceQueryStr),
+							})
+						}
+					}
+					return filters
+				}(),
+			}
+		}
+
+		state.Scp = append(state.Scp, schedulerJobs)
```
Comment (om-shrirao) [inline internal/provider/scheduler/data_source_scheduler.go:223]: same Scp is used here, make it scheduler of job

File: internal/provider/scheduler/data_source_scheduler.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+			Name:        types.StringValue(jobs.Name),
+			Description: types.StringValue(jobs.Description),
+			Operation:   types.StringValue(jobs.Operation),
+			RunAt:       types.StringValue(jobs.RunAt),
+			RunOn:       types.StringValue(jobs.RunOn),
+			Disabled:    types.BoolValue(jobs.Disabled),
+			StartDate:   types.StringValue(jobs.StartDate.Format(time.RFC3339)),
+			EndDate:     types.StringValue(jobs.EndDate.Format(time.RFC3339)),
+		}
+		if jobs.DatabaseBackupParams != nil {
```
Comment (om-shrirao) [inline internal/provider/scheduler/data_source_scheduler.go:181]: will it work for other operation if they are added later

File: internal/provider/scheduler/resource_scheduler.go [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Required:    true,
+				Description: "The name of the job configuration",
+			},
+			"operation": schema.StringAttribute{
+				Required:    true,
+				Description: "The database_backup_params param must be specified for the database_backup operation",
```
Comment (om-shrirao) [inline internal/provider/scheduler/resource_scheduler.go:73]: is this correct description?

File: internal/provider/scheduler/data_source_scheduler.go (hunk 3 above)
Comment (VJ-VijayThales) [reply internal/provider/scheduler/data_source_scheduler.go:114]: yes everything taken

File: internal/provider/scheduler/data_source_scheduler.go (hunk 4 above)
Comment (VJ-VijayThales) [reply internal/provider/scheduler/data_source_scheduler.go:141]: im also thinking the same as the limit is hard code here, if user wants to see more data then we have no option to see more

File: internal/provider/scheduler/resource_scheduler.go (hunk 7 above)
Comment (VJ-VijayThales) [reply internal/provider/scheduler/resource_scheduler.go:73]: its the correct description as we only support database_backup operation. But ill modify it better

File: internal/provider/scheduler/schema_scheduler.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	DoSCP          types.Bool          `tfsdk:"do_scp"`
+	Connection     types.String        `tfsdk:"connection"`
+}
+
+type BackupFilterTFSDK struct {
+	ResourceType  types.String `tfsdk:"resource_type"`
+	ResourceQuery types.String `tfsdk:"resource_query"`
+}
+
+type CreateJobConfigParamsListJSON struct {
```
[the remark this answers is not in the document]
Comment (VJ-VijayThales) [reply internal/provider/scheduler/schema_scheduler.go:76]: yes the list json is different in the tag, though the contents are same, you can see the database backup param json tag is different in listjson as `job_config_params`
for the normal CreateJobConfigParamsJSON it is `database_backup_params`

File: sample-scripts/data-sources/scheduler/README.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
```
[the remark this answers is not in the document]
Comment (VJ-VijayThales) [reply sample-scripts/data-sources/scheduler/README.md:2]: we have it in docs folder also.. and separate one in sample scripts as well

[the remark this answers is not in the document]
Comment (om-shrirao) [reply sample-scripts/data-sources/scheduler/README.md:2]: but i am not able to see any doc folder.

File: internal/provider/scheduler/data_source_scheduler.go (hunk 4 above)
Comment (om-shrirao) [reply internal/provider/scheduler/data_source_scheduler.go:141]: make it -1

File: sample-scripts/scheduler/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+- `updated_at` (String)
+- `uri` (String)
+
+<a id="nestedatt--database_backup_params"></a>
+### Nested Schema for `database_backup_params`
+
+Optional:
+
+- `backup_key` (String) ID of backup key used for encrypting the backup. The default backup key is used if this is not specified.
+- `connection` (String) Name or ID of the SCP connection which stores the details for SCP server.
```
[the remark this answers is not in the document]
Comment (VJ-VijayThales) [reply sample-scripts/scheduler/README.md:55]: this is basically under the struct database param, so i think it itself portrays as it belongs to only database backup params

[the remark this answers is not in the document]
Comment (VJ-VijayThales) [reply sample-scripts/scheduler/README.md:55]: this actually comes under the section `Nested Schema for database_backup_params`, so it itself portrays that it comes under the database backup params

**Review by om-shrirao — APPROVED** [review]

**Pull request merged**